packages/angular_devkit/core/src/jobs/README.md

```diff
@@ -0,0 +1,386 @@
+# Description
```

Comment:
I had the comment below on the original PR on the `Job Extension` section.

I can't find this section anymore. Has job extension overall been removed from this PR?

---

I think this section really needs some examples and more details. Things like a job extending each other, what it means to replace a job, and a replacement example.

Another thing that I don't understand is the usage pattern for jobs that have a superset of the input and outputs that is not. 

Consider I have a JobFoo, with known inputs and outputs. JobBar extends JobFoo, and has additional inputs and outputs.

If I want to transparently replace JobFoo with JobBar, I can only rely on using the known input/output for JobFoo. Thus the additional input/output that JobBar has will always use defaults.

I'd like to understand how this impacts the use cases that Schematics and Architect have for Jobs.

For instance, having a generic `node_install` job that is ultimately replaced by `npm_install` or `yarn_install`, or some other. 

Should the base `node_install` have only the intersection of the input/output of the (known) possible replacements? Or should it have some input/output that we expect to be ignored by some of the replacements?

Replacements are one of the key requirements of Jobs so it's important that this be clear.

Reply:
The way to handle replacements is to either use a dispatcher (recommended), or use a registry and register a different job under the same name (not so recommended).

I'm working to explain `implements` better. If you have any questions, we can schedule a hangout.

Reply:
The choice of whether to use a dispatcher is if the job to use will change at runtime.  If it will be static for the course of execution, the overhead and complexity of a dispatcher would be unnecessary.

Reply:
Actually it would be recommended to have a dispatcher instead because jobs could refer to non-generic jobs. But that's about usage more than API definition.

Reply:
I'd rather `implements` isn't in this PR at all if it's not crystal clear in the readme. 

Like I said before: replacements are one of the key requirements of Jobs so it's important that this be clear.

It's ok if `implements` doesn't exist for the initial implementation. It can be added and reviewed later.

This is already a very big PR. I haven't even started to look at the code because even reviewing the readme has taken this long and there are still questions that I haven't gotten answers to.

Besides, it looks like there are already two ways to handle replacements (dispatcher and re-register) so it doesn't look like it's super urgent to have a third one.

Reply:
`implements` is part of the dispatcher logic.

Reply:
> I'd rather implements isn't in this PR at all if it's not crystal clear in the readme.

You're welcome to make suggestions. It's clear to me just by the name (dispatches to one job).

I'd like to throw the README to Jody to get feedback on the documentation. Maybe I'll do that today.

> Like I said before: replacements are one of the key requirements of Jobs so it's important that this be clear.

I think the confusion comes from the scope of this PR to be as small as possible, while keeping an honest list of things that show usage. I could have removed any mention of replacing a job with another and it would fit the scope. Dispatchers are only in there to show that it's possible to make a dispatcher **without** any change to the API.

The Jobs API is _really_ powerful, in part because of its simplicity. You'll combine jobs and utility functions to break down complex behaviour into manageable code. Without dispatchers, and even without scheduler implementations, this PR would have been the same. Except that it would have been more confusing as these questions would still pop up.

Maybe I should just remove Dispatchers, Registry and NodeModuleScheduler from the README itself. Would that help?

> I haven't even started to look at the code because even reviewing the readme has taken this long and there are still questions that I haven't gotten answers to.

Considering the amount of comment around the code, you need to.

> implements is part of the dispatcher logic.

`implements` is **not** part of the dispatcher logic. It's part of the registration logic. The Registry change the schema to add the additional implementation schemas.

The dispatcher could not even be in this PR and `implements` would be WAI.

Reply:
I'm on board with having the small scope for this PR. I mostly want that the whole "replacements" use case is thoroughly discussed, not just a comment thread in the middle of a feature rich PR.

If there is nothing about replacements/implements in the readme or in the public API that's good enough for me. That way when it needs to be added to the public API we can have the discussion.

> Considering the amount of comment around the code, you need to.

I want to get to the rest but it just sounds like a waste of time when things are still changing just from discussing the design in the readme. We agreed that the readme was enough of a design doc so it makes sense to review the design before the implementation, especially when it's still changing.

It's now the 7th day of comments in the readme for me and there are at least 7 force pushes in this PR. It is impossible for me as a reviewer to track changes. Whenever I want to see what changed in the readme alone I have to re-read it from the start. I am very glad that I didn't review the rest yet because I'd also have to continuously scan it for differences every review, wasting a huge amount of time.

Reply:
Removed `implements` from the Registry interface. We can add it as a feature later.